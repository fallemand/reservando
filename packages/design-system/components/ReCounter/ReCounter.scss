@import "../../styles/variables.scss";

.re-counter {
  font-size: 1em;
  display: flex;
  align-items: center;

  &__cta {
    flex-shrink: 0;
  }

  &__icon {
    font-size: 1em;
    margin: 0.5em 1em;
    flex-shrink: 0;
  }

  &__input {
    margin: 0 0.5em;
    text-align: center;
    flex-grow: 1;

    .re-input__field,
    .re-input__field::-webkit-outer-spin-button,
    .re-input__field::-webkit-inner-spin-button {
      -moz-appearance: textfield;
      -webkit-appearance: none;
      margin: 0;
    }
  }
}
