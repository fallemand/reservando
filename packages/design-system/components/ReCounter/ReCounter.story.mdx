import { Meta, Story, ArgsTable, Canvas, Description } from "@storybook/addon-docs/blocks";
import { ReCounter } from "@reservando/design-system";

<Meta title="Components/Counter" component={ReCounter} />

# ReCounter

<Canvas>
  <Story name="basic">
    {{
      components: { ReCounter },
      data() {
        return {
          value: 0,
        };
      },
      template: `
        <div style="max-width: 200px;">
          <ReCounter v-model="value" />
        </div>
      `,
    }}
  </Story>
</Canvas>

## How to use

```html
<template>
  <ReCounter v-model="value" />
</template>

<script lang="ts">
  import { ReCounter } from "@reservando/design-system";
</script>
```

<ArgsTable of={ReCounter} />

## Examples

#### Using params

Step: `2` | min: `10` | max: `20`

<Canvas>
  <Story name="with params">
    {{
      components: { ReCounter },
      data() {
        return {
          value: 10,
        };
      },
      template: `
        <div style="max-width: 200px;">
          <ReCounter :min="10" :max="20" :step="2" v-model="value" />
        </div>
      `,
    }}
  </Story>
</Canvas>
