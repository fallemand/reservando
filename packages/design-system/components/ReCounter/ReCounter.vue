<template>
  <div class="re-counter">
    <ReButtonRound
      class="re-counter__cta"
      size="small"
      modifier="secondary"
      :disabled="minDisabled"
      icon="minus"
      @click="decrease"
    />
    <ReInput
      ref="input"
      type="number"
      class="re-counter__input"
      :model-value="modelValue"
      :min="min"
      :max="max"
      :step="step"
      @change="inputChange"
    />
    <ReButtonRound
      class="re-counter__cta"
      size="small"
      modifier="secondary"
      :disabled="maxDisabled"
      icon="plus"
      @click="increase"
    />
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import ReButtonRound from "../ReButtonRound/ReButtonRound.vue";
import ReInput from "../ReInput/ReInput.vue";

const ReCounter = defineComponent({
  name: "ReCounter",
  components: {
    ReButtonRound,
    ReInput,
  },
  props: {
    /** Counter value */
    modelValue: {
      required: true,
      type: Number,
    },
    step: {
      type: Number,
      default: 1,
    },
    min: {
      type: Number,
      default: 0,
    },
    max: {
      type: Number,
      default: null,
    },
  },
  computed: {
    minDisabled(): boolean {
      return this.modelValue - this.step < this.min;
    },
    maxDisabled(): boolean {
      if (!this.max) {
        return false;
      }
      return this.modelValue + this.step > this.max;
    },
  },
  created(): void {
    if (this.modelValue < this.min) {
      this.$emit("update:modelValue", this.min);
    }
    if (this.max && this.modelValue > this.max) {
      this.$emit("update:modelValue", this.max);
    }
  },
  methods: {
    lessThanMin(value: number): boolean {
      return value < this.min;
    },
    greaterThanMax(value: number): boolean {
      if (!this.max) {
        return false;
      }
      return value > this.max;
    },
    increase(): void {
      this.$emit("update:modelValue", this.modelValue + this.step);
    },
    decrease(): void {
      this.$emit("update:modelValue", this.modelValue - this.step);
    },
    inputChange(value: string): void {
      let numberValue = Number(value);
      if (this.lessThanMin(value)) {
        numberValue = this.min;
      }
      if (this.greaterThanMax(value)) {
        numberValue = this.max;
      }
      /** Emited when value changed `(value: number)` */
      this.$emit("update:modelValue", numberValue);
    },
  },
});

export default ReCounter;
</script>

<style lang="scss" src="./ReCounter.scss" />
