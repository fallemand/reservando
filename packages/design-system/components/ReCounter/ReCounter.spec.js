import { shallowMount, mount } from "@vue/test-utils";
import ReCounter from "./ReCounter.vue";

const getMinusButton = (wrapper) => wrapper.find(".re-counter__cta:first-child");
const getPlusButton = (wrapper) => wrapper.find(".re-counter__cta:last-child");

describe("ReCounter component", () => {
  const props = {
    modelValue: 0,
  };

  it("should render with default props", () => {
    const wrapper = shallowMount(ReCounter, {
      props,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should initial value be higher or equal than min", () => {
    const wrapper = shallowMount(ReCounter, {
      props: {
        modelValue: 2,
        min: 10,
      },
    });

    expect(wrapper.emitted()["update:modelValue"][0]).toEqual([10]);
  });

  it("should initial value be lower or equal than max", () => {
    const wrapper = shallowMount(ReCounter, {
      props: {
        modelValue: 20,
        max: 10,
      },
    });

    expect(wrapper.emitted()["update:modelValue"][0]).toEqual([10]);
  });

  it("should decrease value by 1", async () => {
    const wrapper = mount(ReCounter, {
      props: {
        modelValue: 2,
      },
    });

    await getMinusButton(wrapper).trigger("click");

    expect(wrapper.emitted()["update:modelValue"][0]).toEqual([1]);
  });

  it("should increase value by 1", async () => {
    const wrapper = mount(ReCounter, {
      props: {
        modelValue: 1,
      },
    });

    await getPlusButton(wrapper).trigger("click");

    expect(wrapper.emitted()["update:modelValue"][0]).toEqual([2]);
  });

  it("should button be disabled when lower border reached.", () => {
    const wrapper = shallowMount(ReCounter, {
      props: {
        min: 10,
        modelValue: 10,
      },
    });

    expect(wrapper.vm.minDisabled).toBe(true);
  });

  it("should button be disabled when higher border reached.", () => {
    const wrapper = shallowMount(ReCounter, {
      props: {
        max: 10,
        modelValue: 10,
      },
    });

    expect(wrapper.vm.maxDisabled).toBe(true);
  });
});
