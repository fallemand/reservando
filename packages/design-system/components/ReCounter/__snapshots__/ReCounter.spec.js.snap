// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReCounter component should render with default props 1`] = `
<div
  class="re-counter"
>
  <re-button-stub
    class="re-counter__cta"
    disabled="true"
    size="small"
  />
  <re-input-stub
    class="re-counter__input"
    min="0"
    modelvalue="0"
    step="1"
    type="number"
  />
  <re-button-stub
    class="re-counter__cta"
    disabled="false"
    size="small"
  />
</div>
`;
